nixos/doc/manual/development/option-types.section.md

```diff
@@ -400,44 +400,55 @@ Composed types are types that take a type as parameter. `listOf
 
 `types.attrsOf` *`t`*
 
-:   An attribute set of where all the values are of *`t`* type. Multiple
-    definitions result in the joined attribute set.
+:   _Shorthand for `types.attrsWith { lazy = false; elemType = t; }`._
 
-    ::: {.note}
-    This type is *strict* in its values, which in turn means attributes
-    cannot depend on other attributes. See `
-           types.lazyAttrsOf` for a lazy version.
-    :::
+    An attribute set of where all the values are of *`t`* type.
+
+    This type is *strict* in its values, which in turn means attributes cannot depend on other attributes.
+    See `lazy` in `attrsWith`.
 
 `types.lazyAttrsOf` *`t`*
 
-:   An attribute set of where all the values are of *`t`* type. Multiple
-    definitions result in the joined attribute set. This is the lazy
-    version of `types.attrsOf
-          `, allowing attributes to depend on each other.
+:   _Shorthand for `types.attrsWith { lazy = true; elemType = t; }`._
 
-    ::: {.warning}
-    This version does not fully support conditional definitions! With an
-    option `foo` of this type and a definition
-    `foo.attr = lib.mkIf false 10`, evaluating `foo ? attr` will return
-    `true` even though it should be false. Accessing the value will then
-    throw an error. For types *`t`* that have an `emptyValue` defined,
-    that value will be returned instead of throwing an error. So if the
-    type of `foo.attr` was `lazyAttrsOf (nullOr int)`, `null` would be
-    returned instead for the same `mkIf false` definition.
-    :::
+    An attribute set of where all the values are of *`t`* type.
+
+    This is the lazy version of `types.attrsOf`, allowing attributes to depend on each other, at the cost of potentially containing more attributes than desirable.
+    See `lazy` in `attrsWith`.
 
 `types.attrsWith` { *`elemType`*, *`lazy`* ? false, *`placeholder`* ? "name" }
 
 :   An attribute set of where all the values are of *`elemType`* type.
 
+    Multiple definitions result in the joined attribute set.
+
     **Parameters**
 
     `elemType` (Required)
     : Specifies the type of the values contained in the attribute set.
 
     `lazy`
-    : Determines whether the attribute set is lazily evaluated. See: `types.lazyAttrsOf`
+    : Determines whether the set of attribute names is evaluated before or after evaluating any attribute values.
+
+      `lazy = true` is a preferable choice in principle, but requires extra effort if the option value is used in aggregate (e.g. `attrNames` or `toJSON`).
+
+      When `lazy`, the attribute names are determined by looking at the definitions without evaluating them.
+      If only a `mkIf false` occurs in an attribute, this may result in an error when the attribute value is accessed.
```

Comment:
```suggestion
      If a `mkIf false` occurs in an attribute, this may result in an error when the attribute value is accessed.
```

Artefact? Or am I misreading the intention behind that sentence?

Reply:
When an option has multiple defs, and only one is `mkIf false`, we're fine.

Technically this should be "if all definitions are `mkIf false`", but that's overly mechanistic IMO.

To make things worse, reading that as a _for all_, we'd think of the `all f [] == true` thing, which doesn't apply here, because the `[]` does not occur in this context.